lib/mr.js

```diff
@@ -10,20 +10,24 @@ MillerRabin.create = function create(rand) {
   return new MillerRabin(rand);
 };
 
-MillerRabin.prototype._rand = function _rand(n) {
+MillerRabin.prototype._randbelow = function _randbelow(n) {
   var len = n.bitLength();
-  var buf = this.rand.generate(Math.ceil(len / 8));
+  var min_bytes = Math.ceil(len / 8);
 
-  // Set low bits
-  buf[0] |= 3;
+  // Generage random bytes until a number less than n is found.
+  // This ensures that 0..n-1 have an equal probability of being selected.
+  do
+    var a = new bn(this.rand.generate(min_bytes));
```

Comment:
Let's keep bit masking from the original function, but the check from your version. It doesn't add any entropy bias to have it and would generally be faster, right?

Reply:
The main thing we need to do is ensure that numbers of the form `x*n - 1, x*n, x*n + 1` are not used as bases. These numbers do not provide valid results when used in the Miller-Rabin test. Although the odds of selecting them is inversely proportional to the value of n, they should not be included.

It should be sufficient to choose a random base that has one less bit than n - 1. My research actually indicates that there is a binary symmetry in the bases. The top half can safely be excluded, because choosing `x` as a base is equivalent to choosing `n - x` as a base. I would be surprised if this hasn't been mentioned in a research paper before, but I don't have a reference on hand. This would still include some of the bases in the top half of the range, but since they are equivalent to those in the bottom half it would effectively be skewing the distribution of the RNG towards the center of the range.

Less skewing of the distribution could be achieved by using a modulo operation to bring the random bytes into the desired range, but it is unclear if the performance of this operation would be better than resampling.

Reply:
Right now it'd have to do extra resampling just because we didn't discard high bits that are not going to be present in the result anyway. Discarding high bits is much much faster than comparing two big numbers twice, not including random number generation and conversion of it to the BN.

Reply:
I think we need some concrete benchmarks. I suspect that there is no significant difference between the resampling strategy and the truncating strategy in the context of the overall run time. Intuitively they should profile much differently, but in the context of the entire algorithm I suspect they are insignificant compared to the expense of modular exponentiation. We should not sacrifice correctness for an insignificant optimization.

Besides, the Miller-Rabin is already a non-deterministic algorithm. It would be nice to have a non-deterministic RNG to keep it company. 😆

Reply:
@deckar01  that's really interesting. I'm been grappling with the idea for a while - I'm writing this function into a smart contract on the Ethereum platform, and the costs of regenerating an expensive random number should the random number be in the top half of the range are too great - but if I can always generate one less bit than n-1 the cost of the random number call becomes deterministic :D

Reply:
> a random base that has one less bit than n - 1 ... would effectively be skewing the distribution of the RNG towards the center of the range.

@splut It does skew the distribution, but I do not have any references for how it affects the error rate of the Miller-Rabin test. I suspect it does not have any significant negative impact on the error rate when testing large numbers. FWIW, we ended up using the non-deterministic algorithm that Python uses, because modular exponentiation is orders of magnitude more expensive.

Reply:
I might test this directly so - it's super important to get right for my application. Thanks